File: frontend/types/industry-collaboration.ts
// Mirrors the `industry_collaborations` table
// (database/migrations/026_industry_collaborations.sql) and
// backend/app/schemas/industry_collaboration.py. Keep all three in sync.
//
// This is NOT a posting entity like IndustryProject/IndustryTraining/
// IndustryWorkshop/IndustryMentorship — it is a bilateral academia-
// industry collaboration proposal/relationship between an INDUSTRY
// account (initiator) and a FACULTY or INSTITUTION account (recipient),
// with its own lifecycle:
// DRAFT -> SENT -> ACCEPTED/REJECTED -> ACTIVE -> COMPLETED/CANCELLED.
//
// Named `industry-collaboration` (not `collaboration`/`collaborations`)
// to avoid colliding with the still-unimplemented, ambiguous generic
// stubs and with 009_collaboration.sql's own broader, unimplemented
// scope — both untouched.

export const COLLABORATION_STATUSES = [
  "DRAFT",
  "SENT",
  "ACCEPTED",
  "REJECTED",
  "ACTIVE",
  "COMPLETED",
  "CANCELLED",
] as const;
export type CollaborationStatus = (typeof COLLABORATION_STATUSES)[number];

export const COLLABORATION_STATUS_LABELS: Record<CollaborationStatus, string> = {
  DRAFT: "Draft",
  SENT: "Sent",
  ACCEPTED: "Accepted",
  REJECTED: "Rejected",
  ACTIVE: "Active",
  COMPLETED: "Completed",
  CANCELLED: "Cancelled",
};

export const RECIPIENT_TYPES = ["FACULTY", "INSTITUTION"] as const;
export type RecipientType = (typeof RECIPIENT_TYPES)[number];

export const RECIPIENT_TYPE_LABELS: Record<RecipientType, string> = {
  FACULTY: "Faculty",
  INSTITUTION: "Institution",
};

export interface IndustryCollaboration {
  id: string;
  industry_id: string;
  recipient_id: string;
  recipient_type: RecipientType;
  title: string;
  description: string;
  status: CollaborationStatus;
  created_at: string | null;
  updated_at: string | null;
  /** Display identity of each party, resolved server-side (migration 029 —
   * `collaboration_counterparty_names()`). `industry_name` is the
   * initiator's company name; `recipient_name` is the Faculty/Institution
   * account's name. Either may be null (migration not yet applied, or no
   * name on file) — fall back to the recipient-type label. */
  industry_name?: string | null;
  recipient_name?: string | null;
}

/** POST body. Always created as DRAFT — no `status`, no `industry_id`,
 * no `recipient_type` (derived server-side from `recipient_id`'s real
 * role). `recipient_id` is normally obtained via the recipient resolver
 * first. */
export interface CollaborationCreate {
  title: string;
  description: string;
  recipient_id: string;
}

/** PUT body — partial, DRAFT-only. `recipient_id` is never editable. */
export interface CollaborationUpdate {
  title?: string;
  description?: string;
}

/** Minimal recipient lookup result — id/role/full_name only, never
 * email/phone/other profile fields. */
export interface RecipientResolution {
  id: string;
  role: RecipientType;
  full_name: string | null;
}

File: backend/app/schemas/industry_collaboration.py
"""Pydantic schemas for Industry collaborations.

Field names and constraints match database/migrations/026_industry_collaborations.sql
exactly (`industry_collaborations`). This is NOT a posting entity like
industry_project/industry_training/industry_workshop/industry_mentorship
(022-025) -- it is a bilateral academia-industry collaboration
proposal/relationship between an INDUSTRY account (initiator) and a
FACULTY or INSTITUTION account (recipient), with its own lifecycle:
DRAFT -> SENT -> ACCEPTED/REJECTED -> ACTIVE -> COMPLETED/CANCELLED.

Named `industry_collaboration` (not `collaboration`/`collaborations`) to
avoid colliding with the still-unimplemented, ambiguous generic stubs
(backend/app/api/collaborations.py, backend/app/schemas/collaboration.py)
and with 009_collaboration.sql's own broader, unimplemented scope.

Ownership (`industry_id`) and lifecycle (`status`) are never accepted from
the client -- `industry_id` is always the authenticated caller, and
`status` only changes through the explicit lifecycle endpoints.
`recipient_type` is never accepted from the client either -- it is
derived server-side (by a database trigger) from the referenced
recipient's real `profiles.role`. `extra="forbid"` on the write models is
what structurally rejects an attempt to smuggle any of these in.
"""

from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

CollaborationStatus = Literal[
    "DRAFT", "SENT", "ACCEPTED", "REJECTED", "ACTIVE", "COMPLETED", "CANCELLED"
]
RecipientType = Literal["FACULTY", "INSTITUTION"]

COLLABORATION_STATUSES: tuple[str, ...] = (
    "DRAFT",
    "SENT",
    "ACCEPTED",
    "REJECTED",
    "ACTIVE",
    "COMPLETED",
    "CANCELLED",
)
RECIPIENT_TYPES: tuple[str, ...] = ("FACULTY", "INSTITUTION")


class CollaborationCreate(BaseModel):
    """POST /api/v1/collaborations. Always created as DRAFT -- there is no
    `status` field here, and no `industry_id`. `recipient_id` must be the
    id of an existing FACULTY or INSTITUTION profile (typically obtained
    via GET /collaborations/recipients/resolve first) -- the database
    trigger rejects anything else and derives `recipient_type` itself, so
    there is no `recipient_type` field here either."""

    model_config = ConfigDict(extra="forbid")

    title: str = Field(min_length=1, max_length=200)
    description: str = Field(min_length=1, max_length=10_000)
    recipient_id: UUID


class CollaborationUpdate(BaseModel):
    """PUT /api/v1/collaborations/{id}. Partial: only the fields actually
    sent are changed. Only permitted while the collaboration is still
    DRAFT -- the service layer rejects this once sent. `recipient_id` is
    immutable after creation (not editable here, ever) and `status`
    cannot be changed here."""

    model_config = ConfigDict(extra="forbid")

    title: str | None = Field(default=None, min_length=1, max_length=200)
    description: str | None = Field(default=None, min_length=1, max_length=10_000)


class IndustryCollaboration(BaseModel):
    id: str
    industry_id: str
    recipient_id: str
    recipient_type: str
    title: str
    description: str
    status: str
    created_at: str | None = None
    updated_at: str | None = None
    # Display identity of each party, resolved server-side via the
    # collaboration_counterparty_names() function (migration 029) -- the
    # read-side counterpart of the recipient resolver. `industry_name` is
    # the initiator's company_name (profiles.full_name fallback);
    # `recipient_name` is the FACULTY/INSTITUTION account's full_name.
    # Null when migration 029 has not been applied yet, or (rarely) when a
    # party has no name on file -- the UI falls back to the recipient-type
    # label. Never accepted from the client; not stored on the table.
    industry_name: str | None = None
    recipient_name: str | None = None


class CollaborationListResponse(BaseModel):
    collaborations: list[IndustryCollaboration]


class RecipientResolution(BaseModel):
    """GET /api/v1/collaborations/recipients/resolve response. Deliberately
    minimal -- id, role, full_name only. No email/phone/other profile
    fields, and only ever populated for a FACULTY or INSTITUTION match."""

    id: str
    role: str
    full_name: str | None = None
